src/global/install.rs

```diff
@@ -444,63 +503,122 @@ pub(crate) async fn sync(
+            install_environment(
+                &specs,
+                &env_name,
+                &parsed_environment,
+                auth_client.clone(),
+                &prefix,
+                config,
+                &gateway,
+            )
+            .await?;
+        }
+
+        expose_executables(
+            &env_name,
+            &parsed_environment,
+            specs.keys().cloned().collect(),
+            &prefix,
+            &bin_dir,
         )
         .await?;
     }
 
     Ok(())
 }
+
+/// Checks if the local environment matches the given specifications.
+///
+/// This function verifies that all the given specifications are present in the
+/// local environment's prefix records and that there are no extra entries in
+/// the prefix records that do not match any of the specifications.
+fn specs_match_local_environment(
+    specs: &IndexMap<PackageName, MatchSpec>,
+    prefix_records: Vec<PrefixRecord>,
+    platform: Option<Platform>,
+) -> bool {
+    // Check whether all specs in the manifest are present in the installed environment
+    let specs_in_manifest_are_present = specs.iter().all(|(name, spec)| {
+        let Some(prefix_record) = prefix_records
+            .iter()
+            .find(|record| record.repodata_record.package_record.name == *name)
+        else {
+            return false;
+        };
+
+        spec.matches(&prefix_record.repodata_record)
+    });
+
+    if !specs_in_manifest_are_present {
+        return false;
+    }
+
+    // Check whether all packages in the installed environment have the correct package
+    let platform_specs_match_env = prefix_records.iter().all(|record| {
+        let Ok(package_platform) =
+            Platform::from_str(&record.repodata_record.package_record.subdir)
+        else {
+            return true;
+        };
+
+        match package_platform {
+            Platform::NoArch => true,
+            p if Some(p) == platform => true,
+            _ => false,
+        }
+    });
+
+    if !platform_specs_match_env {
+        return false;
+    }
+
+    fn prune_dependencies(
+        remaining_prefix_records: Vec<PrefixRecord>,
+        matched_record: &PrefixRecord,
+    ) -> Vec<PrefixRecord> {
+        let dependencies = &matched_record.repodata_record.package_record.depends;
+        dependencies
+            .iter()
+            .fold(remaining_prefix_records, |mut acc, dependency| {
+                let Some(dependency_name) =
+                    MatchSpec::from_str(dependency, ParseStrictness::Lenient)
+                        .ok()
+                        .and_then(|d| d.name)
+                else {
+                    return acc;
+                };
+
+                let Some(index) = acc.iter().position(|record| {
+                    record.repodata_record.package_record.name == dependency_name
+                }) else {
+                    return acc;
+                };
```

Comment:
Instead of extracting the name you can also just use `matches` on the matchspec itself. That way you also figure out if the environment is inconsistent.